The polymer is formed by smaller units which, when triggered, react with each other such that two adjacent units of the same type and opposite polarity are destroyed. Units' types are represented by letters; units' polarity is represented by capitalization. For instance, r and R are units with the same type but opposite polarity, whereas r and s are entirely different types and do not react.

For example:

    In aA, a and A react, leaving nothing behind.
    In abBA, bB destroys itself, leaving aA. As above, this then destroys itself, leaving nothing.
    In abAB, no two adjacent units are of the same type, and so nothing happens.
    In aabAAB, even though aa and AA are of the same type, their polarities match, and so nothing happens.

Now, consider a larger example, `dabAcCaCBAcCcaDA`:

```
dabAcCaCBAcCcaDA  The first 'cC' is removed.
dabAaCBAcCcaDA    This creates 'Aa', which is removed.
dabCBAcCcaDA      Either 'cC' or 'Cc' are removed (the result is the same).
dabCBAcaDA        No further actions can be taken.
```

After all possible reactions, the resulting polymer contains 10 units.

How many units remain after fully reacting the polymer you scanned?

In [2]:
def get_input_from_file(filename):
    with open(filename) as f:
        for line in f:
            yield line

In [36]:
input_line = ""
for line in get_input_from_file("input_5.txt"):
    input_line = line
#     print(line)

In [4]:
def check_letters(a, b):
    if a == b:
        return False
    if str(a).lower() == str(b):
        return True
    elif str(a) == str(b).lower():
        return True
    return False

# check_letters('A', 'A')
# check_letters('a', 'A')
# check_letters('A', 'a')
# check_letters('a', 'a')

In [5]:
import itertools
def pairwise(iterable):
    "s -> (s0,s1), (s1,s2), (s2, s3), ..."
    a, b = itertools.tee(iterable)
    next(b, None)
    return zip(a, b)

iter_pair = pairwise(input_line)

In [22]:
%%time
# print(input_line)
def find_length_of_reduced_polymer(input_line):
    input_line = input_line + "-"
    while len(input_line) > 0:
        reduced_line = ""
        iter_pair = pairwise(input_line)
        for c, n in iter_pair:
            if check_letters(c, n):
                next(iter_pair)
            else:
                reduced_line += c
        input_line = input_line[:len(input_line) - 1]
        if input_line == reduced_line:
            break
        input_line = reduced_line + "-"

    input_line = input_line.strip()
    return len(input_line)

        
print(find_length_of_reduced_polymer(input_line))

11242
CPU times: user 9.64 s, sys: 1.78 ms, total: 9.64 s
Wall time: 9.67 s


## Part two

For example, again using the polymer dabAcCaCBAcCcaDA from above:

    Removing all A/a units produces dbcCCBcCcD. Fully reacting this polymer produces dbCBcD, which has length 6.
    Removing all B/b units produces daAcCaCAcCcaDA. Fully reacting this polymer produces daCAcaDA, which has length 8.
    Removing all C/c units produces dabAaBAaDA. Fully reacting this polymer produces daDA, which has length 4.
    Removing all D/d units produces abAcCaCBAcCcaA. Fully reacting this polymer produces abCBAc, which has length 6.

In this example, removing all C/c units was best, producing the answer 4.

What is the length of the shortest polymer you can produce by removing all units of exactly one type and fully reacting the result?

In [39]:
from string import ascii_uppercase as up_case_string
# print(len(input_line))
character_n_length_of_reduced_polymer = {}
for c in list(up_case_string):
    duplicate_line = input_line
    duplicate_line = duplicate_line.replace(c, "")
    duplicate_line = duplicate_line.replace(c.lower(), "")
    character_n_length_of_reduced_polymer[c] = find_length_of_reduced_polymer(duplicate_line)
#     print(len(duplicate_line))

# print(character_n_length_of_reduced_polymer)

{'A': 10768, 'B': 10808, 'C': 10750, 'D': 10842, 'E': 10800, 'F': 10784, 'G': 10718, 'H': 10806, 'I': 10786, 'J': 10822, 'K': 10826, 'L': 10770, 'M': 10812, 'N': 10820, 'O': 10752, 'P': 10796, 'Q': 10776, 'R': 10832, 'S': 10816, 'T': 5492, 'U': 10804, 'V': 10826, 'W': 10754, 'X': 10836, 'Y': 10752, 'Z': 10768}


Thus we can say that by removing `T`, we can get shortest polymer of length `5492`

In [42]:
# print(character_n_length_of_reduced_polymer)
for k, v in character_n_length_of_reduced_polymer.items():
    if v == min(character_n_length_of_reduced_polymer.values()):
        print(k, v)

T 5492
